### ISSUE TITLE
NUnit3TestAdapter 3.15 not running tests with custom TestCaseSource

### ISSUE THREAD

**Issue by MattKeenum** [nunit/nunit3-vs-adapter#650]
NUnit 3.12.0
NUnit3TestAdapter 3.15.0
Visual Studio 2019 Version 16.1.3
.NET Framework 4.6.1

```
namespace NUnitTest
{
   public class UnitTest1
    {
        public class DataProvider<T> where T : TestData, new()
        {
            public static IEnumerable<TestCaseData> GetData()
            {
                string[] names = new string[] { "Matthew", "Kelly", "Shawn" };
                char[] genders = new char[] { 'M', 'F', 'M' };
                int[] ages = new int[] { 35, 34, 30 };

                for (int i = 0; i < names.Length; i++)
                {
                    TestData testData = new TestData()
                    {
                        Name = names[i],
                        Gender = genders[i],
                        Age = ages[i]
                    };

                    TestCaseData tcd = new TestCaseData(testData);
                    tcd.SetName("{m} [" + testData.Name + " - " + testData.Gender + " - " + testData.Age + "]");

                    yield return tcd;
                }
            }
        }


        [Test]
        [TestCaseSource(typeof(DataProvider<TestData>), "GetData")]
        public void Persons(TestData test)
        {
            Assert.IsTrue(test.Name == "Matthew");
        }
    }
}
```

The above code correctly generates 3 test cases that have the following names in the Test Explorer:
```
Persons [Kelly - F - 34]
Persons [Matthew - M - 35]
Persons [Shawn - M - 30]
```
However, when I attempt to run/debug any of those test cases, they disappear from the Test Explorer, and nothing happens. A simple rebuild (or another clean and build) brings them back, but they still cannot be executed as they just disappear again.

Using the debugger, the custom data 

If I downgrade to NUnit3TestAdapater 3.14, everything works as expected, I get 2 failed and 1 passed.

**Comment by OsirisTerje** [comment]
Thanks  @MattKeenum  !   Based on this case, in addition to the other, I am pretty sure we need to rollback 3.15.  That performance work is unfortunately not finished.

**Comment by turcunicusor** [comment]
@MattKeenum How can I downgrade to NUnit3TestAdapater 3.14, without using NuGet?

**Comment by turcunicusor** [comment]
I found [https://github.com/nunit/nunit3-vs-adapter/releases](url) all the Visual Studio test Adapters. Just uninstall current version of NUnit, download the proper .vsix file, and then install it.

**Comment by OsirisTerje** [comment]
@jnm2 Should SetName be deprecated?  And have a comment added it to that warns about this kind of stuff, and pointing to the SetArgDisplayNames ?  Ref documentation at https://github.com/nunit/docs/wiki/TestCaseData which doesnt mention SetArgDisplayNames, nor warns about this issue with SetName.  I see there are multiple NUnit framework issues around the same thing.

**Comment by MattKeenum** [comment]
@jnm2 @OsirisTerje Thanks for the workaround. Unfortunately we have some complex ways we build tests and the Set Names worked great since it gave full control over how the names were set. Any chance we can get that feature with SetArgsDisplayNames?

**Comment by jnm2** [comment]
@MattKeenum Maybe, can you explain more? SetArgDisplayNames only allows you to customize the display text of each parameter, but it uses the standard namespace, type name, method name, parentheses, and commas. Are some of those things you need to override?

**Comment by MattKeenum** [comment]
@jnm2 Sure. Our company has over 500 applications that all need automated testing. Since we are mostly a C# shop, we chose NUnit as the testing framework of choice. As each team developed their own framework to write automated tests, we realized that every team was solving the same problems in different ways. 

We wrote a generic wrapper around NUnit and distribute it as a NuGet package. This lets us solve the common problems once, and easily allows team members to move from team to team without having to learn a different framework.

Apart of this, we heavily customize the TestCaseSource to be able to create tests from different data sources, like Excel, CSV, Databases, etc. We then duplicate these tests changing slight configurations, like which environment to test, where the tests are executed (Local, Selenium Grid, Sauce Labs, etc.). If the team is using mobile devices, it allows the tests to be duplicated for different devices. This allows us to build once and run the tests we need to when approaching the CI/CD pipeline. 

Unfortunately, I am prohibited by my company to upload pictures, but here are some example test names:

[Local : Dev : Win : Chrome] - [Card] - [MyMethod1] - [MyDataSourceTestName]
[Grid : Dev : Win : Chrome] - [Card] - [MyMethod1] - [MyDataSourceTestName]
[Sauce Labs : QA : Mac : Safari] - [Card] - [MyMethod2] - [MyDataSourceTestName]

Having full control over how the names are presented, makes it much easier to look through the Test Explorer for the tests the end user wants to run. 

I'm open to ideas. We aren't 100% sold on SetNames. Just realize that any changes we make to the way the Test Names are displayed in the Test Explorer, affects 100s of people using the framework. So we want to be consistent in how we display these names to reduce confusion among the teams.

**Comment by jnm2** [comment]
@MattKeenum That makes sense. Thanks for the detail. SetArgDisplayNames is by design not the way to go for your use case. We'll continue to look at what it's going to take to get SetNames interacting well with the prefilter.

**Comment by OsirisTerje** [comment]
@MattKeenum @turcunicusor  There is now a beta version of the fix in https://www.myget.org/feed/nunit/package/nuget/NUnit3TestAdapter/3.15.1-dev-01134  .  Would appreciate if you checked it.

**Comment by OsirisTerje** [comment]
@jnm2   Will you add a new issue for the framework ?   I think this one can be closed,  right ?

**Comment by jnm2** [comment]
Should we transfer this issue there or start a new one?

**Comment by turcunicusor** [comment]
@OsirisTerje I checked the beta version fix. Now it behaves as expected. I want to mention that I am using SetUpFixture with OneTimeSetUp, and custom TestCaseSource. Thank's a lot!

**Comment by OsirisTerje** [comment]
@turcunicusor  Thanks for the check-up :-)  !

@jnm2  I think it would be good to start a new one.  Then this one here will serve as documentation for the issue in the adapter.  I've linked to this one in the release notes, the "top" item #651 and in the Tips and Tricks page https://github.com/nunit/docs/wiki/Tips-And-Tricks#prefilter

**Comment by OsirisTerje** [comment]
3.15.1 hotfix released now.

**Comment by MattKeenum** [comment]
@OsirisTerje I can confirm that the fix is working for 3.15.1. Thanks everyone!

**Comment by yuriyoliin** [comment]
An Issue is reproduced again.
NUnit3TestAdapter 3.15.1/3.16/3.16.1/3.17 not running tests with custom TestCaseSource
